libbeat/monitoring/report/elasticsearch/client.go

```diff
@@ -80,15 +80,42 @@ func (c *publishClient) Close() error {
 func (c *publishClient) Publish(batch publisher.Batch) error {
 	events := batch.Events()
 	bulk := make([]interface{}, 0, 2*len(events))
+
+	var params map[string]string
 	for _, event := range events {
+
+		// Extract time
+		t, err := event.Content.Meta.GetValue("type")
+		if err != nil {
+			logp.Err("Type not available in monitoring reported. Please report this error: %s", err)
+			continue
+		}
+
+		params = map[string]string{}
+		// Copy params
+		for k, v := range c.params {
+			params[k] = v
+		}
+		// Extract potential additional params
+		p, err := event.Content.Meta.GetValue("params")
+		if err == nil {
+			p2, ok := p.(map[string]string)
+			if ok {
+				for k, v := range p2 {
+					params[k] = v
+				}
+			}
+		}
```

Comment:
how about us copying/allocating contents if required only:

```
var params, userParams, clientParams map[string]string
clientParams = c.params
if p, err := event.Content.Meta.GetValue("params"); err == nil {
  if tmp, ok := p.(map[string]string); ok {
    userParams = tmp
  }
}

switch {
  case len(userParams) == 0:
    params = clientParams
  case len(clientParams) == 0:
    params = userParams
  default:
    params = make(map[string]string, len(userParams) + len(clientParams))
    for k, v := range clientParams {
      params[k] = v
    }
    for k, v := range userParams {
      params[k] = v
    }
}
```

Reply:
Is c.params and params via 'meta' always configured? If not, we can reduce some copying here:

```
eventParams := func(meta common.MapStr) map[string]string {
  tmp, err := event.Content.Meta.GetValue("params")
  if err != nil {
    return nil
  }
  p, ok := tmp.(map[string]string)
  if !ok {
    return nil
  }
  return p
}

cpyKV := func(to map[string]string, from map[string]string) {
  for k, v := range from {
    to[k] = v
  }
}

params = c.params
if userParams := eventParams(event.Content.Meta); len(userParams) > 0 {
  if len(c.params) > 0 {
    params = make(map[string]string, len(params) + len(userParams))
    cpyKV(params, c.params)
    cpyKV(params, userParams)
  } else {
    params = userParams
  }
}
```

Reply:
As each event contains the interval param in meta and it's required at the moment, I think we have to reprocess it every time.

What do you mean by we `loose` params? As it's reprocessed every time and we don't modify c.params, all "base" params should always be there?